src/commands/dev/lib/dev-server.ts

```diff
@@ -47,31 +49,99 @@ export default class DevServer {
   public output: Output;
   public env: EnvConfig;
   public buildEnv: EnvConfig;
+  public files: BuilderInputs;
 
+  private cachedNowJson: NowConfig | null;
   private server: http.Server;
   private stopping: boolean;
   private buildMatches: Map<string, BuildMatch>;
   private inProgressBuilds: Map<string, Promise<void>>;
   private originalEnv: EnvConfig;
+  private nsfw?: nsfw.Watcher;
 
   constructor(cwd: string, options: DevServerOptions) {
     this.cwd = cwd;
     this.output = options.output;
     this.env = {};
     this.buildEnv = {};
+    this.files = {};
 
+    this.cachedNowJson = null;
     this.server = http.createServer(this.devServerHandler);
     this.stopping = false;
     this.buildMatches = new Map();
     this.inProgressBuilds = new Map();
     this.originalEnv = { ...process.env };
   }
 
-  async getProjectFiles(): Promise<BuilderInputs> {
-    // TODO: use the file watcher to keep the files list up-to-date
-    // incrementally, instead of re-globbing the filesystem every time
-    const files = await collectProjectFiles('**', this.cwd);
-    return files;
+  async handleFilesystemEvents(events: nsfw.Event[]): Promise<void> {
+    const filesChanged: Set<string> = new Set();
+
+    // First, update the `files` mapping of source files
+    for (const event of events) {
+      // TODO: for some reason the type inference isn't working, hence the casting
+      if (event.action === nsfw.actions.CREATED) {
```

Comment:
You might try switch(event.action) instead

Reply:
Tried it, doesn't fix the type inference unfortunately, so I'm still not sure what the issue is there.